### PR TITLE
Azure backup hunting queries

### PR DESCRIPTION
**Pull request by thmcelro**
**Create AzureBackupCeased.yaml**

This query can be used during threat hunts to identify potentially malicious tampering with backup policies used by Microsoft Recovery Services Vaults/Azure Backup. The query specifically identifies instances where a backup has stopped being performed on a resource that was previously being backed up. The query executes over a 30 days training window to identify existing backups, it will then perform detection over the most recent 3 days. Additionally, in instances where the backed up resource is a VM, the query will join with virtual machine deletion events to determine if backups ceased due to resource deletion.

Query has been tested in several production environments with good results.

**Create AzureBackupsRetentionPolicyReduction.yaml**

This hunting query will detect anomalies where a backup policy modification has taken place that reduces backup reliability. The query considers three possible reductions: A reduction in the frequency of days for backups (e.g. backup 5 days a week to only 1 day a week), a reduction in the retention period for a backup (e.g. retain for 30 days reduced to 60 days) or a reduction in the frequency of backups per-day (e.g. backup twice a day reduced to once). A policy modification resulting in a reduction of any of these parameters may indicate malicious tampering with the intended to erode backup integrity.

Query has been tested in several production environments with good results. A level of noise is expected, FP events will be generated when a policy is legitimately updated.

### CODE REVIEW

File: Hunting Queries/AzureActivity/AzureBackupsCeased.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    dataTypes:
+      - AzureActivity
+tactics:
+  - Impact
+relevantTechniques:
+  - T1485
+  - T1490
+query: |
+  // Total time range for the query, includes detection and training data
+  let timeRange = 60d;
```
Comment (shainw) [inline Hunting Queries/AzureActivity/AzureBackupsCeased.yaml:20]: For hunting queries time ranges, we need to support the features of the hunting blade which requires these be in a certain format.  See - https://github.com/Azure/Azure-Sentinel/blob/master/Hunting%20Queries/AzureActivity/Rare_Custom_Script_Extension.yaml - for example usage.

  let starttime = todatetime('{{StartTimeISO}}');
  let endtime = todatetime('{{EndTimeISO}}');
  let Lookback = starttime - 14d;

File: Hunting Queries/AzureActivity/AzureBackupsCeased.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      | where TimeGenerated > ago(timeRange)
+      | where OperationNameValue =~ "Microsoft.Compute/virtualMachines/delete"
+      | extend ResourceId = tolower(tostring(iff(isempty(ResourceId), _ResourceId, ResourceId)))
+      | extend Deleted = iff(ActivityStatusValue has_any ('Success', 'Succeeded'), 1, 0)
+      | summarize max(Caller), max(CallerIpAddress), min(TimeGenerated), max(TimeGenerated), max(Deleted) by CorrelationId,ResourceId, tolower(ResourceGroup)
+      | extend Resource = tolower(tostring(split(ResourceId, "/")[-1]))
+  ) on Resource
+  | project Resource, ResourceId, LastSeenBackup, DeletedDate=min_TimeGenerated, ResourceDeleted=tobool(max_Deleted), CorrelationId, PreviousBackups=list_TimeGenerated, TotalBackups=dcount_TimeGenerated
+  // Optionally filter on retention periods that were reduced but the resource was not deleted
+  //| where ResourceDeleted == False
```
Comment (shainw) [inline Hunting Queries/AzureActivity/AzureBackupsCeased.yaml:64]: We want to do entity mappings in hunting queries to support future features.  Can you add in the entity mapping for ResourceId.

File: Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  | extend Days = split(BackupDaysOfTheWeek, ","), Times = split(BackupTimes, ",")
+  | extend Days = array_length(Days), Times = array_length(Times)
+  | order by PolicyUniqueId, TimeGenerated asc
+  | extend nxtPolicy = next(PolicyUniqueId), nxtDays = next(Days), nxtTimes = next(Times), nxtMonthlyRetentionDuration = next(MonthlyRetentionDuration), nxtTimeGenerated=next(TimeGenerated)
+  | where PolicyUniqueId == nxtPolicy and (nxtDays < Days or nxtTimes < Times or nxtMonthlyRetentionDuration < MonthlyRetentionDuration)
+  | project PolicyUniqueId, PolicyName, TimeGenerated, AnomalyTimeGenerated = nxtTimeGenerated, DayChange = iff(nxtDays < Days, strcat(Days, " > ", nxtDays), "No Change"), TimeChange = iff(nxtTimes < Times, strcat(Times, " > ", nxtTimes), "No Change"), MonthlyRetentionDurationChange = iff(nxtMonthlyRetentionDuration < MonthlyRetentionDuration, strcat(MonthlyRetentionDuration, " > ", nxtMonthlyRetentionDuration), "No Change"),  DayAnomalyFlag = iff(nxtDays < Days, 1, 0),  TimeAnomalyFlag = iff(nxtTimes < Times, 1, 0), MonthlyRetentionDurationAnomalyFlag = iff(nxtMonthlyRetentionDuration < MonthlyRetentionDuration, 1, 0)
```
Comment (shainw) [inline Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml:27]: We want to do entity mappings in hunting queries to support future features.  Can you add in the entity mapping for ResourceId.  Looks like you will need to bring it thru in your Project here also.

File: Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      - AddonAzureBackupPolicy
+tactics:
+  - Impact
+relevantTechniques:
+  - T1485
+  - T1490
+query: |
+  AddonAzureBackupPolicy
+  | where OperationName =~ "Policy"
+  | project TimeGenerated, PolicyUniqueId, PolicyName, BackupFrequency, MonthlyRetentionDuration, BackupDaysOfTheWeek, BackupTimes, VaultUniqueId
```
Comment (shainw) [inline Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml:21]: We should bring thru ResourceId.

**Review by shainw — CHANGES_REQUESTED** [review]
In general, query and yaml look good, couple of minor changes.

**Comment by v-sabiraj** [comment]
Hey @thmcelro, We have resolved the issue for PR Validations failure and request you to please update your PR branch with the latest changes from **Azure:master** so that the PR Validations can be executed.... Thanks!!!

**Comment by v-sabiraj** [comment]
Hey @thmcelro , please update the requested changes.. thanks

File: Hunting Queries/AzureActivity/AzureBackupsCeased.yaml (hunk 1 above)
Comment (thmcelro) [reply Hunting Queries/AzureActivity/AzureBackupsCeased.yaml]: Added

File: Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml (hunk 3 above)
Comment (thmcelro) [reply Hunting Queries/AzureActivity/AzureBackupsRetentionPolicyReduction.yaml]: Updated

**Comment by v-sabiraj** [comment]
Hey @shainw, will you please check and approve the requested changes, thanks.

File: Hunting Queries/AzureActivity/AzureBackupsCeased.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  AzureActivity
+  | where TimeGenerated between(starttime .. endtime)
+  | where ResourceProviderValue =~ "MICROSOFT.RECOVERYSERVICES"
+  // Extract the resource name (in this case the VM name)
+  | extend Resource = tostring(Properties_d.["entity Name"])
+  | project TimeGenerated, Resource, Claims_d, OperationNameValue, ResourceGroup, Caller
+  | where isnotempty(Resource);
+  // Process backups events in the training window
+  rawData
+  | where TimeGenerated between (starttime) .. end)
```
Comment (aprakash13) [inline Hunting Queries/AzureActivity/AzureBackupsCeased.yaml:38]: We might need to add a ( at the start for the query to work.

**Review by aprakash13 — CHANGES_REQUESTED** [review]
Running the query gives an error because of a missing ( . Please add this . Thanks.

**Comment by vmanojreddy** [comment]
Hey @thmcelro  , please update the requested changes.. thanks

**Comment by shainw** [comment]
Closing for now as no response from poster.

**Pull request closed without merging**